src/logger.ts

```diff
@@ -1,29 +1,94 @@
-"use strict";
+import * as winston from "winston";
+import * as Transport from "winston-transport";
 
-const winston = require("winston");
+export type LogLevel =
+  | "error"
+  | "warn"
+  | "help"
+  | "data"
+  | "info"
+  | "debug"
+  | "prompt"
+  | "http"
+  | "verbose"
+  | "input"
+  | "silly";
 
-function expandErrors(logger) {
-  const oldLogFunc = logger.log;
-  logger.log = function (...logArgs) {
-    const args = logArgs.slice(0);
-    if (args.length >= 2 && args[1] instanceof Error) {
-      args[1] = args[1].stack;
+// Extend the Winston log methods to support error signatures
+export interface LogMethod extends winston.LogMethod {
+  (level: LogLevel, err: Error, ...meta: any[]): Logger;
+}
+
+export interface LeveledLogMessage extends winston.LeveledLogMethod {
+  // We use empty log messages to create newlines
+  (): Logger;
+
+  // We transform Errors to strings dynamically
+  (err: Error, ...meta: any[]): Logger;
+}
+
+export interface Logger {
+  log: LogMethod;
+
+  error: LeveledLogMessage;
+  warn: LeveledLogMessage;
+  help: LeveledLogMessage;
+  data: LeveledLogMessage;
+  info: LeveledLogMessage;
+  debug: LeveledLogMessage;
+  prompt: LeveledLogMessage;
+  http: LeveledLogMessage;
+  verbose: LeveledLogMessage;
+  input: LeveledLogMessage;
+  silly: LeveledLogMessage;
+
+  add(transport: Transport): Logger;
+  remove(transport: Transport): Logger;
+}
+
+function expandErrors(logger: winston.Logger): winston.Logger {
+  const oldLogFunc: winston.LogMethod = logger.log.bind(logger);
+  const newLogFunc: winston.LogMethod = function (
+    levelOrEntry: string | winston.LogEntry,
+    message?: string | Error,
+    ...meta: any[]
+  ): winston.Logger {
+    if (message && message instanceof Error) {
+      message = message.stack || message.name;
+      return oldLogFunc(levelOrEntry as string, message, ...meta);
     }
-    return oldLogFunc.apply(this, args);
+    // Overloads are weird in TypeScript. This method works so long as the original
+    // function isn't checking arguments.length.
+    return oldLogFunc(levelOrEntry as string, message as string, ...meta);
   };
+  logger.log = newLogFunc;
   return logger;
 }
 
-const logger = expandErrors(winston.createLogger());
+function annotateDebugLines(logger: winston.Logger): winston.Logger {
+  const debug: winston.LeveledLogMethod = logger.debug.bind(logger);
+  const newDebug: winston.LeveledLogMethod = function (
+    message: string | any | object,
```

Comment:
Why do we need `any | object` here? It seems a bit redundant to me & I don't see any special casing for Object here.